### ISSUE TITLE
compileall fails due to bad demo code

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6908]
|<img alt="jlynch's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jlynch reported|
|-|-|
|Trac ID|trac#6908|
|Type|defect|
|Created|2014-01-17 22:50:51Z|

If I attempt to run compileall in the base of the twisted package directory:

```
python -m compileall .
```

I get three SyntaxErrors in demo code within the docs directory. In particular:

```
SyntaxError: ('invalid syntax', ('./docs/projects/historic/2003/pycon/deferex/deferex-bad-adding.py', 6, 1, '...\n'))
SyntaxError: ('invalid syntax', ('./docs/projects/historic/2003/pycon/deferex/deferex-listing0.py', 8, 9, '        ...\n'))
SyntaxError: ('invalid syntax', ('./docs/projects/historic/2003/pycon/deferex/deferex-listing2.py', 6, 1, '...\n'))
```

It makes sense that these are not valid python code as they are just demos, but since the files are still .py files, it makes sense that they should compile.  

I have attached a patch as per my understanding of the [wiki](http://twistedmatrix.com/trac/wiki/TwistedDevelopment#SubmittingaPatch) that makes these files valid python . I did not include a [news](http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles) file because I am not sure which topfiles directory is appropriate as this is a change to the docs subdirectory. 

Attachments:

* [fix-compileall-6908.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a62/a6242d53e14fedfa7d062c22376f7112791c41cc/62b5c002c4edcb04e462acb9381c233bb7fbd76f.patch) (1492 bytes) - added by jlynch on 2014-01-17 22:52:40Z - Patch that makes demo code valid python

<details><summary>Searchable metadata</summary>

```
trac-id__6908 6908
type__defect defect
reporter__jlynch jlynch
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__compileall compileall
time__1389999051000000 1389999051000000
changetime__1603710227718867 1603710227718867
version__None None
owner__jlynch jlynch

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to jlynch|
|-|-|

Why does it matter what compileall does to our documentation?

**Comment by twisted-trac** [comment]
|<img alt="jlynch's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|jlynch commented|
|-|-|

It matters because production python shops might want to run compileall before they actually ship their pycs to their production environments.

Some reasons one might want to do this:

1. Additional line of defense against invalid code making it to production (in addition to pyflakes, flake8, etc ...).

2. Performance concerns when the code paths first get hit.

3. Security. Usually the user executing the production python code is allowed read permission but not write permission to the filesystem so that even if someone compromised that user they would be less able to impact the machine. This means that you must generate the pyc's before deployment.

You would be right to assume that in a production system you might be using virtualenv and proper python packaging, which resolves this issue because we would never see a .py from a doc directory, but I happen to be at one of the older python shops that built quite a lot of infrastructure before proper packaging and did a lot of cross repository dependency management with git submodules. Basically we are maintaining a fork right now with this issue fixed and I figured it would be better to just get this contributed back upstream since it is a quick and harmless change to "to the right thing".


Also just in general if you have python files in a python package they should probably be valid python, even if they are in an example directory.

**Comment by twisted-trac** [comment]
|<img alt="jlynch's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|jlynch commented|
|-|-|

Bumping this back to review because I think it is a quick easy fix that makes all .py files properly compile in the repository.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

I'm still not inclined to accept this.

  * The historic documentation is there as a record of what was presented at a conference.  Changing it distorts this record.
  * There is no test coverage that demonstrates that "all .py files properly compile".  Nor is there any development policy that requires this to be the case.
  * Twisted is in the "twisted" directory in the repository.  If you want to compile all of Twisted, you only have to compile that directory, not all of the docs.  As a bonus, this will be faster and use less disk space.

I'll let someone else agree or disagree and claim the review though.

**Comment by twisted-trac** [comment]
|<img alt="jlynch's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|jlynch commented|
|-|-|

Ok, I mean it is perfectly reasonable to say that the twisted package is just the directories provided by the twisted python package (which is just the twisted subdirectory and not the twisted directory). It also seems reasonable to say that if you have a .py file in your repository, it should probably be valid python.  Like I said above, you are right that using proper python deps would be the "right way" on our end (and save disk sapce, and be faster because you can use wheels, etc ...); unfortunately it's taking a long time to move our large code base towards using virtualenv and proper packaging.

I'm submitting this patch because it is 6 lines and means we don't have to maintain a fork of twisted. To address your concerns, would the probability of this patch getting accepted go up if I got the blessing of glyph, who I believe gave the talk, and/or if I included a test case that asserts compileall succeeds in the base directory?

To be honest I figured this would just be an easy patch since it's in a doc directory and not impacting the api in any way. I know that this is impacting development in at least one large python code base that uses git submodules (so a full checkout of the twisted repo), and I don't believe it is implausible to assume there are others out there.

**Comment by twisted-trac** [comment]
|<img alt="jlynch's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|jlynch set status to `closed`|
|-|-|

As per the discussion over at #twisted-dev I'm going to try to work -x into our deploy process instead of this patch.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Thanks for working through this with us.  Even though we didn't merge anything, it's always nice to have users bring us their problems rather than silently raging at Twisted :-).  Glad you got it resolved!

**Comment by twisted-trac** [comment]
|[<img alt="graingert's avatar" src="https://avatars.githubusercontent.com/u/413772?s=50" width="50" height="50">](https://github.com/graingert)|@graingert set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="graingert's avatar" src="https://avatars.githubusercontent.com/u/413772?s=50" width="50" height="50">](https://github.com/graingert)<a name="note_9"></a>|@graingert set status to `closed`|
|-|-|

this was fixed in 3d177ca10f76c47c57d48d9a271a0caea0eea625
